File: go/greeter/greeter.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type InputHandler struct{}

func (inputHandler InputHandler) getUserName() string {
  fmt.Print("Enter your name: ")
  reader := bufio.NewReader(os.Stdin)
  userInput, _ := reader.ReadString('\n')

  name := strings.TrimSpace(userInput)

  return name
}

type Greeter struct {
  username string
}

func NewGreeter(name string) *Greeter {
  return &Greeter{username: name}
}

func (greeter *Greeter) Greet() {
  fmt.Printf("Hello, %s!\n", greeter.username)
  fmt.Println("Greetings from Go!")
}

func main() {
  name := InputHandler{}.getUserName()
  greeter := NewGreeter(name)

  greeter.Greet()
}

File: c/greeter/src/greeter.c
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char* GetUserName() {
  char userinput[100];
  printf("Enter your name: ");
  if (fgets(userinput, 100, stdin) == NULL) {
    return NULL;
  }

  char* name = strdup(userinput);
  name[strlen(name)-1] = '\0';

  name += strspn(name, " \t\n\r");
  size_t len = strlen(name);
  while (len > 0 && isspace((unsigned char)name[len - 1])) {
    name[--len] = '\0';
  }

  if (strlen(name) == 0) {
    strcpy(name, "Stranger");
  }

  return name;
}

typedef struct {
  char* username;
} Greeter;

void Greeter_Greet(Greeter* greeter) {
  printf("Hello, %s!\n", greeter->username);
  puts("Greetings from C!");
}

int main() {
  char* name = GetUserName();
  Greeter greeter;
  greeter.username = name;

  Greeter_Greet(&greeter);

  return 0;
}
